File: src/specter/secure_bus.py
"""Signed in-process message bus + replay protection.

Threat model coverage:
- Tampering: any modification to envelope fields invalidates signature.
- Spoofing: receiver verifies signature against claimed sender's roster key.
- Replay: per-sender nonce must be strictly monotonic.

Phase 03 swaps the in-process bus for SROS2 transport using the same envelope
contract (ADR 0001 + 0003).
"""

import json
import time
from collections.abc import Callable
from dataclasses import dataclass, field

from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePublicKey

from .crypto import Keypair, public_from_bytes, verify
from .interfaces import MessageBus

WIRE_VERSION = 1


class VerificationError(Exception):
    """Envelope failed signature, roster, version, or replay check."""


@dataclass(frozen=True)
class Envelope:
    version: int
    sender_id: str
    nonce: int
    timestamp_ns: int
    kind: str
    payload: bytes
    signature: bytes


def _signed_blob(
    version: int, sender_id: str, nonce: int, timestamp_ns: int, kind: str, payload: bytes
) -> bytes:
    return json.dumps(
        {
            "version": version,
            "sender_id": sender_id,
            "nonce": nonce,
            "timestamp_ns": timestamp_ns,
            "kind": kind,
            "payload": payload.hex(),
        },
        sort_keys=True,
        separators=(",", ":"),
    ).encode()


def envelope_to_wire(env: Envelope) -> bytes:
    return json.dumps(
        {
            "version": env.version,
            "sender_id": env.sender_id,
            "nonce": env.nonce,
            "timestamp_ns": env.timestamp_ns,
            "kind": env.kind,
            "payload": env.payload.hex(),
            "signature": env.signature.hex(),
        },
        sort_keys=True,
        separators=(",", ":"),
    ).encode()


def envelope_from_wire(data: bytes) -> Envelope:
    blob = json.loads(data)
    return Envelope(
        version=blob["version"],
        sender_id=blob["sender_id"],
        nonce=blob["nonce"],
        timestamp_ns=blob["timestamp_ns"],
        kind=blob["kind"],
        payload=bytes.fromhex(blob["payload"]),
        signature=bytes.fromhex(blob["signature"]),
    )


@dataclass
class Roster:
    """agent_id → public key. Distributed at swarm formation."""

    keys: dict[str, EllipticCurvePublicKey] = field(default_factory=dict)

    def add(self, agent_id: str, public_bytes: bytes) -> None:
        self.keys[agent_id] = public_from_bytes(public_bytes)


@dataclass
class Identity:
    """An agent's signing identity + outbound nonce counter."""

    agent_id: str
    keypair: Keypair
    _nonce: int = 0

    def next_nonce(self) -> int:
        self._nonce += 1
        return self._nonce

    def seal(self, kind: str, payload: bytes, timestamp_ns: int | None = None) -> Envelope:
        nonce = self.next_nonce()
        ts = timestamp_ns if timestamp_ns is not None else time.time_ns()
        blob = _signed_blob(WIRE_VERSION, self.agent_id, nonce, ts, kind, payload)
        return Envelope(WIRE_VERSION, self.agent_id, nonce, ts, kind, payload, self.keypair.sign(blob))


@dataclass
class ReplayWindow:
    """Strict-monotonic per-sender nonce tracker."""

    last_seen: dict[str, int] = field(default_factory=dict)

    def accept(self, sender_id: str, nonce: int) -> bool:
        if nonce <= self.last_seen.get(sender_id, 0):
            return False
        self.last_seen[sender_id] = nonce
        return True


def open_envelope(env: Envelope, roster: Roster, replay: ReplayWindow) -> bytes:
    if env.version != WIRE_VERSION:
        raise VerificationError(f"unsupported version {env.version}")
    pub = roster.keys.get(env.sender_id)
    if pub is None:
        raise VerificationError(f"unknown sender {env.sender_id}")
    blob = _signed_blob(env.version, env.sender_id, env.nonce, env.timestamp_ns, env.kind, env.payload)
    if not verify(pub, blob, env.signature):
        raise VerificationError(f"bad signature from {env.sender_id}")
    if not replay.accept(env.sender_id, env.nonce):
        raise VerificationError(f"replay from {env.sender_id} nonce={env.nonce}")
    return env.payload


class InProcessBus(MessageBus):
    def __init__(self) -> None:
        self._subs: dict[str, list[Callable[[bytes], None]]] = {}

    def publish(self, topic: str, payload: bytes) -> None:
        for handler in self._subs.get(topic, []):
            handler(payload)

    def subscribe(self, topic: str, handler: Callable[[bytes], None]) -> None:
        self._subs.setdefault(topic, []).append(handler)

File: ui/packages/sim-core/src/secureBus.ts
/**
 * Signed in-process bus + replay protection — TypeScript parity to
 * `src/specter/secure_bus.py`. Used by `runSecureScenario` (see scenario.ts)
 * to round-trip every Observation/PoseReport through seal → wire → openEnvelope
 * → evaluator, so envelope-level attacks (swap_key, replay_storm,
 * forged_envelope) exercise the same code paths as the Python eval harness.
 *
 * Threat model coverage matches Python:
 *   tampering   — any modification to envelope fields invalidates signature
 *   spoofing    — receiver verifies signature against claimed sender's roster key
 *   replay      — per-sender nonce must be strictly monotonic
 */

import { canonicalJsonBytes } from "./canonicalJson";
import {
  envelopeFromWire,
  envelopeToWire,
  openEnvelope,
  seal,
  VerificationError,
  categorizeVerificationError,
  type Envelope,
  type Identity,
  type Roster,
  type RejectCategory,
  type ReplayWindow,
} from "./envelope";
import { generateKeypair, publicKeyFromPrivate } from "./crypto";
import type { ContactReport, Observation, PoseReport } from "./messages";
import { KIND_CONTACT_REPORT, KIND_OBSERVATION, KIND_POSE } from "./messages";

// Payload codec — canonical JSON of the Observation/PoseReport fields. Matches
// `src/specter/messages.py:encode` exactly: bigint timestamps serialized as
// JSON numbers without precision loss (bigint integers fit in JSON until 2^53;
// our 50-tick scenarios at 1e9 ns/s stay well under that cap).

export function encodeObservation(obs: Observation): Uint8Array {
  return canonicalJsonBytes({
    bearing_rad: obs.bearing_rad,
    observer_id: obs.observer_id,
    range_m: obs.range_m,
    subject_id: obs.subject_id,
    timestamp_ns: obs.timestamp_ns,
  });
}

export function encodePoseReport(p: PoseReport): Uint8Array {
  return canonicalJsonBytes({
    agent_id: p.agent_id,
    theta: p.theta,
    timestamp_ns: p.timestamp_ns,
    x: p.x,
    y: p.y,
  });
}

export function decodeObservation(payload: Uint8Array): Observation {
  const blob = JSON.parse(new TextDecoder().decode(payload)) as Record<string, unknown>;
  return {
    observer_id: blob.observer_id as string,
    subject_id: blob.subject_id as string,
    range_m: blob.range_m as number,
    bearing_rad: blob.bearing_rad as number,
    timestamp_ns: BigInt(blob.timestamp_ns as number | string),
  };
}

export function decodePoseReport(payload: Uint8Array): PoseReport {
  const blob = JSON.parse(new TextDecoder().decode(payload)) as Record<string, unknown>;
  return {
    agent_id: blob.agent_id as string,
    x: blob.x as number,
    y: blob.y as number,
    theta: blob.theta as number,
    timestamp_ns: BigInt(blob.timestamp_ns as number | string),
  };
}

export function encodeContactReport(cr: ContactReport): Uint8Array {
  return canonicalJsonBytes({
    contact_id: cr.contact_id,
    kind: cr.kind,
    reporter_id: cr.reporter_id,
    timestamp_ns: cr.timestamp_ns,
    x: cr.x,
    y: cr.y,
  });
}

export function decodeContactReport(payload: Uint8Array): ContactReport {
  const blob = JSON.parse(new TextDecoder().decode(payload)) as Record<string, unknown>;
  return {
    reporter_id: blob.reporter_id as string,
    contact_id: blob.contact_id as string,
    kind: blob.kind as string,
    x: blob.x as number,
    y: blob.y as number,
    timestamp_ns: BigInt(blob.timestamp_ns as number | string),
  };
}

/** Owned identities + receiver-side roster + per-sender replay window.
 * Constructed once per scenario run; the receiver's openEnvelope is called
 * for every published envelope. */
export class SecureBus {
  readonly identities = new Map<string, Identity>();
  readonly roster: Roster = new Map();
  readonly replay: ReplayWindow = new Map();

  /** Mint a keypair, install in roster, return the Identity that seals envelopes. */
  registerAgent(agentId: string): Identity {
    const kp = generateKeypair();
    const id: Identity = { agentId, privateKey: kp.privateKey, nonce: 0n };
    this.identities.set(agentId, id);
    this.roster.set(agentId, kp.publicKey);
    return id;
  }

  /** Mint an off-roster keypair for a forged emitter. Identity returned can
   * seal envelopes with a valid signature, but the roster never learns the
   * public key — openEnvelope rejects with "unknown sender". */
  registerForeignEmitter(emitterId: string): Identity {
    const kp = generateKeypair();
    const id: Identity = { agentId: emitterId, privateKey: kp.privateKey, nonce: 0n };
    this.identities.set(emitterId, id);
    return id;
  }

  /** Rotate an agent's signing keypair without updating the roster — the
   * swap_key attack. All subsequent envelopes from this agent fail signature
   * verification on the receive side. */
  rotateKeypairSecretly(agentId: string): void {
    const id = this.identities.get(agentId);
    if (!id) throw new Error(`no identity for ${agentId}`);
    const kp = generateKeypair();
    id.privateKey = kp.privateKey;
    // Roster is NOT updated — that's the point.
  }

  /** Re-publish a previously captured envelope. Replay window catches it on
   * the second-or-later publish (nonce no longer strictly monotonic). */
  publishCaptured(wire: Uint8Array): { env: Envelope } {
    return { env: envelopeFromWire(wire) };
  }
}

/** Attempt to verify a freshly-sealed envelope. Returns `{ payload }` on
 * success or `{ error, category }` on rejection. Pure wrapper around
 * openEnvelope that buckets the failure category. */
export function tryOpen(
  env: Envelope,
  roster: Roster,
  replay: ReplayWindow,
): { ok: true; payload: Uint8Array } | { ok: false; category: RejectCategory; message: string } {
  try {
    const payload = openEnvelope(env, roster, replay);
    return { ok: true, payload };
  } catch (e) {
    if (e instanceof VerificationError) {
      return { ok: false, category: categorizeVerificationError(e.message), message: e.message };
    }
    throw e;
  }
}

// Convenience helpers for round-tripping a sealed envelope (seal → wire →
// open). Used by tests and scenario.ts envelope mode.

export function sealAndOpenObservation(
  identity: Identity,
  obs: Observation,
  bus: SecureBus,
): { env: Envelope; result: ReturnType<typeof tryOpen> } {
  const env = seal(identity, KIND_OBSERVATION, encodeObservation(obs), obs.timestamp_ns);
  return { env, result: tryOpen(env, bus.roster, bus.replay) };
}

export function sealAndOpenPose(
  identity: Identity,
  p: PoseReport,
  bus: SecureBus,
): { env: Envelope; result: ReturnType<typeof tryOpen> } {
  const env = seal(identity, KIND_POSE, encodePoseReport(p), p.timestamp_ns);
  return { env, result: tryOpen(env, bus.roster, bus.replay) };
}

// Re-exports for convenience
export { envelopeToWire, envelopeFromWire, openEnvelope, seal, VerificationError, categorizeVerificationError };
export type { Envelope, Roster, ReplayWindow, Identity, RejectCategory };
